File: esports-backend/src/main/java/com/esports/controller/PaymentController.java
package com.esports.controller;

import com.esports.dto.ApiResponse;
import com.esports.dto.PaymentHistoryDTO;
import com.esports.entity.MilestonePayment;
import com.esports.service.PaymentService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.NoSuchElementException;

@RestController
@RequestMapping("/api/payments")
@CrossOrigin(origins = "http://localhost:5173")
public class PaymentController {

    @Autowired
    private PaymentService paymentService;

    @GetMapping("/history")
    public ResponseEntity<?> getPaymentHistory(Authentication auth) {
        try {
            List<PaymentHistoryDTO> history = paymentService.getPaymentHistory(auth.getName());
            return ResponseEntity.ok(history);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(ApiResponse.error("User not found"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error("Failed to get payment history"));
        }
    }

    @GetMapping("/invoice/{projectId}")
    public ResponseEntity<?> generateInvoice(@PathVariable Long projectId, Authentication auth) {
        try {
            Map<String, Object> invoice = paymentService.generateInvoice(auth.getName(), projectId);
            return ResponseEntity.ok(invoice);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(ApiResponse.error("Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(ApiResponse.error("Access denied"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error("Failed to generate invoice"));
        }
    }
    
    @PostMapping("/milestone/{projectId}/release")
    @PreAuthorize("hasRole('CREATOR') or hasRole('ADMIN')")
    public ResponseEntity<?> releaseMilestonePayment(@PathVariable Long projectId, @RequestBody Map<String, Object> data, Authentication auth) {
        try {
            Map<String, Object> result = paymentService.releaseMilestonePayment(auth.getName(), projectId, data);
            return ResponseEntity.ok(result);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(ApiResponse.error("Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(ApiResponse.error("Access denied"));
        } catch (IllegalArgumentException e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error("Failed to release milestone payment"));
        }
    }
    
    @GetMapping("/milestones/{projectId}")
    public ResponseEntity<?> getMilestonePayments(@PathVariable Long projectId, Authentication auth) {
        try {
            List<MilestonePayment> milestones = paymentService.getMilestonePayments(auth.getName(), projectId);
            return ResponseEntity.ok(milestones);
        } catch (NoSuchElementException e) {
            return ResponseEntity.status(404).body(ApiResponse.error("Project not found"));
        } catch (SecurityException e) {
            return ResponseEntity.status(403).body(ApiResponse.error("Access denied"));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error("Failed to get milestone payments"));
        }
    }
}

File: Upkart1/Controllers/PaymentController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;
using Upkart1.DTO;
using Upkart1.Entities;
using Upkart1.Services;

namespace Upkart1.Controllers
{
    [ApiController]
    [Route("api/payments")]
    public class PaymentController : ControllerBase
    {
        private readonly IPaymentService _paymentService;

        public PaymentController(IPaymentService paymentService)
        {
            _paymentService = paymentService;
        }

        [HttpGet("history")]
        public IActionResult GetPaymentHistory()
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var history = _paymentService.GetPaymentHistory(email);
                return Ok(history);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(ApiResponse<object>.ErrorResponse("User not found"));
            }
            catch (Exception)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResponse("Failed to get payment history"));
            }
        }

        [HttpGet("invoice/{projectId}")]
        public IActionResult GenerateInvoice(long projectId)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var invoice = _paymentService.GenerateInvoice(email, projectId);
                return Ok(invoice);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(ApiResponse<object>.ErrorResponse("Project not found"));
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, ApiResponse<object>.ErrorResponse("Access denied"));
            }
            catch (Exception)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResponse("Failed to generate invoice"));
            }
        }

        [HttpPost("milestone/{projectId}/release")]
        [Authorize(Roles = "CREATOR,ADMIN")]
        public IActionResult ReleaseMilestonePayment(long projectId, [FromBody] Dictionary<string, object> data)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var result = _paymentService.ReleaseMilestonePayment(email, projectId, data);
                return Ok(result);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(ApiResponse<object>.ErrorResponse("Project not found"));
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, ApiResponse<object>.ErrorResponse("Access denied"));
            }
            catch (ArgumentException ex)
            {
                return BadRequest(ApiResponse<object>.ErrorResponse(ex.Message));
            }
            catch (Exception)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResponse("Failed to release milestone payment"));
            }
        }

        [HttpGet("milestones/{projectId}")]
        public IActionResult GetMilestonePayments(long projectId)
        {
            try
            {
                var email = User.FindFirst(ClaimTypes.Email)?.Value;
                var milestones = _paymentService.GetMilestonePayments(email, projectId);
                return Ok(milestones);
            }
            catch (KeyNotFoundException)
            {
                return NotFound(ApiResponse<object>.ErrorResponse("Project not found"));
            }
            catch (UnauthorizedAccessException)
            {
                return StatusCode(403, ApiResponse<object>.ErrorResponse("Access denied"));
            }
            catch (Exception)
            {
                return StatusCode(500, ApiResponse<object>.ErrorResponse("Failed to get milestone payments"));
            }
        }
    }
}
